File: apps/rh/src/lib/integracao.ts
import 'server-only'
import { SignJWT, jwtVerify } from 'jose'

/**
 * Autenticação entre os módulos — sistema falando com sistema, não pessoa.
 *
 * Reaproveita o AUTH_SECRET que os módulos já precisam compartilhar para o login valer nos
 * dois. Não é preguiça: qualquer segredo novo seria mais uma variável para alguém esquecer
 * de copiar no computador novo, e o modo de falhar seria "a integração parou e ninguém
 * sabe por quê", meses depois. Se o AUTH_SECRET estiver errado, o login já não funciona —
 * o erro aparece no primeiro minuto, não no primeiro EPI entregue.
 *
 * O token é de vida curta e traz quem chamou. Não substitui a sessão do usuário: ele diz
 * "este pedido veio do Almoxarifado", não "este pedido é do fulano".
 */
const EMISSOR_VALIDO = ['estoque', 'rh', 'painel-locacao', 'alojamentos', 'portal', 'programacao'] as const
export type ModuloEmissor = (typeof EMISSOR_VALIDO)[number]

const VALIDADE_SEGUNDOS = 60

function segredo() {
  const s = process.env.AUTH_SECRET
  if (!s || s.length < 32) {
    throw new Error(
      'AUTH_SECRET ausente ou curto demais. Os módulos precisam do MESMO valor para a ' +
        'integração e o login compartilhado funcionarem.',
    )
  }
  return new TextEncoder().encode(s)
}

/** Assina um token para chamar outro módulo. Vida curta: ele viaja em rede local, mas um token eterno vazado seria eterno. */
export async function assinarTokenIntegracao(origem: ModuloEmissor): Promise<string> {
  return new SignJWT({ origem, tipo: 'integracao' })
    .setProtectedHeader({ alg: 'HS256' })
    .setIssuedAt()
    .setExpirationTime(`${VALIDADE_SEGUNDOS}s`)
    .sign(segredo())
}

export type OrigemChamada = { origem: ModuloEmissor }

/**
 * Confere o token de quem está chamando. Devolve null em qualquer problema — assinatura
 * inválida, expirado, ou um token de SESSÃO tentando passar por token de integração.
 *
 * A checagem de `tipo` importa: sem ela, o cookie de sessão de qualquer usuário logado
 * serviria como credencial de máquina, e um endereço pensado para o almoxarifado passaria a
 * aceitar chamada de qualquer pessoa com uma conta.
 */
export async function verificarTokenIntegracao(authorization: string | null): Promise<OrigemChamada | null> {
  const token = authorization?.replace(/^Bearer\s+/i, '').trim()
  if (!token) return null

  try {
    const { payload } = await jwtVerify(token, segredo())
    if (payload.tipo !== 'integracao') return null
    const origem = String(payload.origem) as ModuloEmissor
    if (!EMISSOR_VALIDO.includes(origem)) return null
    return { origem }
  } catch {
    return null
  }
}

File: migracao-go/internal/services/integracao/integracao.go
// Package integracao assina e verifica o token de máquina usado nas chamadas HTTP entre
// módulos (ex.: Almoxarifado -> RH pra ficha de EPI) — não confunde com o cookie de sessão
// de usuário. Espelha lib/integracao.ts: mesmo AUTH_SECRET, HS256, claims {origem, tipo}.
package integracao

import (
	"errors"
	"strings"
	"time"

	"github.com/golang-jwt/jwt/v5"
)

const validadeSegundos = 60

var EmissoresValidos = map[string]bool{
	"estoque": true, "rh": true, "painel-locacao": true, "alojamentos": true, "portal": true,
}

type claims struct {
	Origem string `json:"origem"`
	Tipo   string `json:"tipo"`
	jwt.RegisteredClaims
}

type Servico struct{ segredo []byte }

func NovoServico(segredo string) (*Servico, error) {
	if len(segredo) < 32 {
		return nil, errors.New("AUTH_SECRET ausente ou curto demais para assinar token de integração")
	}
	return &Servico{segredo: []byte(segredo)}, nil
}

// Assinar emite um token de vida curta identificando o módulo de origem — não é sessão de
// usuário, é "este pedido veio do Almoxarifado".
func (s *Servico) Assinar(origem string) (string, error) {
	agora := time.Now()
	c := claims{
		Origem: origem, Tipo: "integracao",
		RegisteredClaims: jwt.RegisteredClaims{
			IssuedAt:  jwt.NewNumericDate(agora),
			ExpiresAt: jwt.NewNumericDate(agora.Add(validadeSegundos * time.Second)),
		},
	}
	return jwt.NewWithClaims(jwt.SigningMethodHS256, c).SignedString(s.segredo)
}

// Verificar confere o cabeçalho Authorization: Bearer ... — devolve a origem se válido, ou
// "", false em qualquer problema (assinatura inválida, expirado, ou um token de SESSÃO de
// usuário tentando passar por token de integração).
func (s *Servico) Verificar(autorizacao string) (origem string, ok bool) {
	token := strings.TrimSpace(strings.TrimPrefix(autorizacao, "Bearer "))
	if token == "" {
		return "", false
	}
	var c claims
	_, err := jwt.ParseWithClaims(token, &c, func(t *jwt.Token) (interface{}, error) {
		return s.segredo, nil
	})
	if err != nil || c.Tipo != "integracao" || !EmissoresValidos[c.Origem] {
		return "", false
	}
	return c.Origem, true
}
